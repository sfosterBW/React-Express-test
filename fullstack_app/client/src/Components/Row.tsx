import React, { FC } from 'react'
import { IFruit } from '../utils/interfaces'

interface Props {
  fruit: IFruit
  handleEdit: () => void
  handleRemove: () => void
  openModal: () => void
}

const Row: FC<Props> = ({ fruit, handleEdit, handleRemove, openModal }) => {
  return (
    <tr>
      <td>
        <p>{fruit._id}</p>
      </td>
      <td>
        <label htmlFor={`${fruit._id}`}>{fruit.name}</label>
      </td>
      <td>
        <p>{fruit.best.toString()}</p>
      </td>
      <td>
        <input
          checked={fruit.best}
          name={`${fruit._id}`}
          onChange={handleEdit}
          type="checkbox"
          value={fruit._id} />
      </td>
      <td>
        <button name="modal" onClick={openModal} value={`${fruit._id}`}>
          Update
        </button>
      </td>
      <td>
        <button name="remove" onClick={handleRemove} value={`${fruit._id}`}>
          Remove
          </button>
      </td>
    </tr>
  )
}

export default Row
