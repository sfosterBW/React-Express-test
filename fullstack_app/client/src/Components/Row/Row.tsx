import React, { FC, useState } from 'react'
import { useDispatch } from 'react-redux'
import { IFruit } from '../../utils/interfaces'
import fruitService from '../../utils/api'
import { openModal, removeFruit, updateFruit } from '../../utils/actions'
import styles from './Row.module.css'

interface Props {
  active?: boolean
  fruit: IFruit
}

const Row: FC<Props> = ({ active = false, fruit }) => {

  const [activeToggle, setActiveToggle] = useState<boolean>(active)
  const dispatch = useDispatch()

  const handleEdit = (fruit: IFruit) => {
    const newFruit = fruit
    newFruit.best = !newFruit.best
    fruitService.updateFruit(newFruit).then(res => dispatch(updateFruit(res)))
  }

  const handleRemove = (fruit: IFruit) => {
    if(fruit._id){
      fruitService.deleteFruit(fruit._id).then((res) => dispatch(removeFruit(res)))
    } else {
      console.log("Fruit undefined", fruit)
    }
  }

  return (
    <div className={styles.row}>
      <div className={styles.header} onClick={() => setActiveToggle(!activeToggle)}>
        <div className={styles.section}>
          <h3 className={styles.subtitle}>{fruit.name}</h3>
        </div>
      </div>
      <div className={`${styles.body} ${activeToggle ? styles.active : undefined}`}>
        <div className={styles.section}>
          <h3 className={styles.subtitle}>Best</h3>
          <input
            checked={fruit.best}
            className={styles.button}
            id={`${fruit._id}`}
            name={`${fruit._id}`}
            onChange={() => handleEdit(fruit)}
            type="checkbox"
            value={fruit._id} />
          <label htmlFor={`${fruit._id}`}>{fruit.best.toString()}</label>
        </div>
        <div className={styles.section}>
          <h3 className={styles.subtitle}>Edit</h3>
          <button
            className={styles.button}
            name="modal"
            onClick={() => dispatch(openModal(fruit))}>
            Update
          </button>
          <button
            className={styles.button}
            name="remove"
            onClick={() => handleRemove(fruit)}
            value={`${fruit._id}`}>
            Remove
          </button>
        </div>
      </div>
    </div>
  )
}

export default Row
