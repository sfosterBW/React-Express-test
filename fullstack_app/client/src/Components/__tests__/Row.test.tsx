
import React from 'react'
import { mount } from 'enzyme'
import renderer from 'react-test-renderer'

import Row from '../Row'

describe('the Row component', () => {

  const fruit = {_id: 1, name:"apple", best:false}
  const mockEditFunction = jest.fn()
  const mockRemoveFunction = jest.fn()
  const mockOpenModalFunction = jest.fn()
  const rowComponent = <Row
    fruit={fruit}
    handleEdit={mockEditFunction()}
    handleRemove={mockRemoveFunction()}
    openModal={mockOpenModalFunction()} />
  const row = mount(rowComponent)

  it('renders with the correct structure', () => {
    expect(row).toBeDefined()
    expect(row.find('tr')).toHaveLength(1)
    expect(row.find('td')).toHaveLength(6)
    expect(row.find('p')).toHaveLength(2)
    expect(row.find('label')).toHaveLength(1)
    expect(row.find('label').text()).toEqual(fruit.name)
    expect(row.find('label').props().htmlFor).toEqual(String(fruit._id))
    expect(row.find('input')).toHaveLength(1)
    expect(row.find('input').props().name).toEqual(String(fruit._id))
    expect(row.find('input').props().type).toEqual('checkbox')
    expect(row.find('button')).toHaveLength(2)
    expect(row.find('button').at(0).props().name).toEqual('modal')
    expect(row.find('button').at(1).props().name).toEqual('remove')
  })

  it('functions as expected',() => {
    row.find('button').at(0).simulate('click')
    row.find('button').at(1).simulate('click')
    expect(mockOpenModalFunction).toHaveBeenCalledTimes(1)
    expect(mockRemoveFunction).toHaveBeenCalledTimes(1)
    row.find('input').simulate('change')
    expect(mockEditFunction).toHaveBeenCalledTimes(1)
  })

  it('renders the same as last time', () => {
    const tree = renderer
      .create(rowComponent)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
