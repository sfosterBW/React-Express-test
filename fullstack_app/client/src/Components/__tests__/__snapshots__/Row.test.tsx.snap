// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the Row component renders the same as last time 1`] = `
<tr>
  <td>
    <p>
      1
    </p>
  </td>
  <td>
    <label
      htmlFor="1"
    >
      apple
    </label>
  </td>
  <td>
    <p>
      false
    </p>
  </td>
  <td>
    <input
      checked={false}
      name="1"
      type="checkbox"
      value={1}
    />
  </td>
  <td>
    <button
      name="modal"
      value="1"
    >
      Update
    </button>
  </td>
  <td>
    <button
      name="remove"
      value="1"
    >
      Remove
    </button>
  </td>
</tr>
`;
